### PR TITLE
initial commit

### PR DESCRIPTION
**Pull request by siordache**
Hi @moditect/moditect-dev, I would love your feedback on this initial implementation.

### CODE REVIEW

**Comment by gunnarmorling** [comment]
@siordache Thanks a lot! I'm going to look into this in the course of the week, just from a very quick look into the Gradle files of the integration tests, this looks super-promising!

@aalmiray, you might like this PR, too :)

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+if: tag IS blank
+
+language: java
+
+sudo: false
+
+before_install: wget https://raw.githubusercontent.com/sormuras/bach/master/install-jdk.sh
```
Comment (aalmiray) [inline .travis.yml]: Why do we have to check for the latest EA builds instead of using standard `oraclejdk8`?

File: build.gradle [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    id 'groovy'
+    id "com.gradle.plugin-publish" version "0.10.0"
+    id "com.github.hierynomus.license" version "0.14.0"
+}
+
+group = 'org.moditect'
+version = '1.0.0-SNAPSHOT'
+
+repositories {
+    mavenCentral()
```
Comment (aalmiray) [inline build.gradle]: Defining `jcenter()` is enough as it contains all entries from `mavenCentral()`.

File: build.gradle [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+sourceCompatibility = 1.8
+
+def defaultEncoding = 'UTF-8'
+[compileJava, compileTestJava]*.options*.encoding = defaultEncoding
+
+[compileGroovy, compileTestGroovy]*.options*.encoding = defaultEncoding
+[compileGroovy, compileTestGroovy]*.groovyOptions*.encoding = defaultEncoding
+
+sourceSets.main.java.srcDirs = []
```
Comment (aalmiray) [inline build.gradle]: These two lines can be deleted

File: build.gradle [hunk 4]
```diff
@@ -0,0 +1,98 @@
+plugins {
+    id 'java-gradle-plugin'
+    id 'groovy'
+    id "com.gradle.plugin-publish" version "0.10.0"
+    id "com.github.hierynomus.license" version "0.14.0"
+}
+
+group = 'org.moditect'
```
Comment (aalmiray) [inline build.gradle]: Move these two properties to `gradle.properties`.

File: build.gradle [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+repositories {
+    mavenCentral()
+    jcenter()
+    gradlePluginPortal()
+    flatDir {
+        dirs "${rootProject.projectDir}/lib"
+    }
+}
+
+sourceCompatibility = 1.8
```
Comment (aalmiray) [inline build.gradle:14]: Move to `gradle.properties`.

File: build.gradle [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+sourceSets.main.java.srcDirs = []
+sourceSets.main.groovy.srcDirs = ["src/main/java", "src/main/groovy"]
+
+configurations {
+    compile {transitive = true}
+}
+
+ext {
+    aetherVersion = '1.1.0'
```
Comment (aalmiray) [inline build.gradle]: Move to `gradle.properties`.

File: build.gradle [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+[compileJava, compileTestJava]*.options*.encoding = defaultEncoding
+
+[compileGroovy, compileTestGroovy]*.options*.encoding = defaultEncoding
+[compileGroovy, compileTestGroovy]*.groovyOptions*.encoding = defaultEncoding
+
+sourceSets.main.java.srcDirs = []
+sourceSets.main.groovy.srcDirs = ["src/main/java", "src/main/groovy"]
+
+configurations {
+    compile {transitive = true}
```
Comment (aalmiray) [inline build.gradle]: Transitive dependencies should be resolved by default, aren't they? Why is this needed here?

File: build.gradle [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+group = 'org.moditect'
+version = '1.0.0-SNAPSHOT'
+
+repositories {
+    mavenCentral()
+    jcenter()
+    gradlePluginPortal()
+    flatDir {
+        dirs "${rootProject.projectDir}/lib"
```
Comment (aalmiray) [inline build.gradle]: Setup a composite build https://docs.gradle.org/current/userguide/composite_builds.html to avoid locally built snapshots

File: src/main/groovy/org/moditect/gradleplugin/ModitectPlugin.groovy [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    final static String EXTENSION_NAME = 'moditect'
+    final static String CREATE_RUNTIME_TASK_NAME = 'createRuntimeImage'
+    final static String GENERATE_MODULE_INFO_TASK_NAME = 'generateModuleInfo'
+    final static String ADD_MAIN_MODULE_INFO_TASK_NAME = 'addMainModuleInfo'
+    final static String ADD_DEPENDENCIES_MODULE_INFO_TASK_NAME = 'addDependenciesModuleInfo'
+    final static String FULL_CONFIGURATION_NAME = 'moditectFull'
+
+    @Override
+    void apply(Project project) {
+        project.configurations.create(FULL_CONFIGURATION_NAME)
```
Comment (aalmiray) [inline src/main/groovy/org/moditect/gradleplugin/ModitectPlugin.groovy]: Use `maybeCreate` instead of `create`

File: src/main/groovy/org/moditect/gradleplugin/ModitectPlugin.groovy [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+        def ext = project.extensions.create(EXTENSION_NAME, ModitectExtension, project)
+        def addMainModuleInfoTask = project.tasks.create(ADD_MAIN_MODULE_INFO_TASK_NAME, AddMainModuleInfoTask)
```
Comment (aalmiray) [inline src/main/groovy/org/moditect/gradleplugin/ModitectPlugin.groovy:41]: Configure `group` and `description` properties for each task.

File: src/main/groovy/org/moditect/gradleplugin/ModitectExtension.groovy [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+            }
+            descriptors.add(new ArtifactDescriptor(info.group, info.name, inputJar))
+        }
+        LOGGER.info "assignedNamesByModules: ${assignedNamesByModules.collect {'\n\t' + it}}"
+        LOGGER.info "artifactDescriptors: ${descriptors.collect {'\n\t' + it}}"
+        new ArtifactsInfo(assignedNamesByModules, descriptors)
+    }
+
+
+    void addMainModuleInfoData(Closure closure) {
```
Comment (aalmiray) [inline src/main/groovy/org/moditect/gradleplugin/ModitectExtension.groovy]: Use `Action` instead of `Closure`. This makes the extension friendly to Kotlin DSL.

File: src/main/groovy/org/moditect/gradleplugin/add/AbstractAddModuleInfoTask.groovy [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+import org.moditect.gradleplugin.Util
+import org.moditect.gradleplugin.add.model.AbstractModuleConfiguration
+import org.moditect.gradleplugin.add.model.ModuleConfiguration
+import org.moditect.gradleplugin.common.ModuleId
+import org.moditect.model.GeneratedModuleInfo
+
+import static org.moditect.gradleplugin.Util.createDirectoryProperty
+
+@CompileStatic
+abstract class AbstractAddModuleInfoTask extends DefaultTask {
```
Comment (aalmiray) [inline src/main/groovy/org/moditect/gradleplugin/add/AbstractAddModuleInfoTask.groovy:41]: Consider making this a `@CacheableTask`.

File: src/main/groovy/org/moditect/gradleplugin/add/AddDependenciesModuleInfoTask.groovy [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+        }
+    }
+
+    AddDependenciesModuleInfoTask() {
+        description = 'Adds module descriptors to existing JAR files'
+        modules = project.objects.property(ModuleList)
+        modules.set(new ModuleList(project))
+    }
+
+    void modules(Closure modulesClosure) {
```
Comment (aalmiray) [inline src/main/groovy/org/moditect/gradleplugin/add/AddDependenciesModuleInfoTask.groovy]: Use `Action` instead of  `Closure`.

File: src/main/groovy/org/moditect/gradleplugin/add/AddMainModuleInfoTask.groovy [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+
+    @Input @Optional
+    final Property<MainModuleConfiguration> mainModule
+
+    AddMainModuleInfoTask() {
+        description = 'Adds a module descriptor to the project JAR'
+        mainModule = project.objects.property(MainModuleConfiguration)
+    }
+
+    void module(Closure closure) {
```
Comment (aalmiray) [inline src/main/groovy/org/moditect/gradleplugin/add/AddMainModuleInfoTask.groovy:43]: Use `Action` instead of `Closure`.

File: src/main/groovy/org/moditect/gradleplugin/add/model/AbstractModuleConfiguration.groovy [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+    abstract String getShortName()
+    abstract File getInputJar()
+    abstract String getVersion()
+    abstract Set<ModuleId> getOptionalDependencies()
+
+    AbstractModuleConfiguration(Project project) {
+        this.project = project
+    }
+
+    void moduleInfo(Closure configureClosure) {
```
Comment (aalmiray) [inline src/main/groovy/org/moditect/gradleplugin/add/model/AbstractModuleConfiguration.groovy]: Use `Action` instead of `Closure`.

File: src/main/groovy/org/moditect/gradleplugin/generate/GenerateModuleInfoTask.groovy [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+        outputDirectory = createDirectoryProperty(project)
+        outputDirectory.set(project.file("$project.buildDir/generated-sources/modules"))
+
+        modules = project.objects.property(ModuleList)
+        modules.set(new ModuleList(project))
+
+        jdepsExtraArgs = project.objects.listProperty(String)
+    }
+
+    void modules(Closure modulesClosure) {
```
Comment (aalmiray) [inline src/main/groovy/org/moditect/gradleplugin/generate/GenerateModuleInfoTask.groovy]: Use `Action` instead of `Closure`.

File: src/main/groovy/org/moditect/gradleplugin/image/CreateRuntimeImageTask.groovy [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+                new ModitectLog()
+        ).run();
+    }
+
+    private Path getJModsDir() {
+        String javaHome = jdkHome.present ? jdkHome.get() : System.getProperty('java.home')
+        new File(javaHome).toPath().resolve("jmods")
+    }
+
+    void launcher(Closure closure) {
```
Comment (aalmiray) [inline src/main/groovy/org/moditect/gradleplugin/image/CreateRuntimeImageTask.groovy:130]: Use `Action` instead of `Closure`.

File: src/main/resources/META-INF/gradle-plugins/org.moditect.gradleplugin.properties [hunk 18]
```diff
@@ -0,0 +1 @@
+implementation-class=org.moditect.gradleplugin.ModitectPlugin
```
Comment (aalmiray) [inline src/main/resources/META-INF/gradle-plugins/org.moditect.gradleplugin.properties]: This file can be removed as it's autogenerated by the `java-gradle-plugin`.

File: build.gradle (hunk 8 above)
Comment (aalmiray) [reply build.gradle]: Huh just remember that moditect core is built with Maven so composite build is not an option.

**Comment by siordache** [comment]
Thanks for your thorough review, @aalmiray ! I will implement the changes in the next few days.

File: build.gradle (hunk 8 above)
Comment (siordache) [reply build.gradle]: I will fetch the moditect artifact from jitpack.

File: src/main/groovy/org/moditect/gradleplugin/add/AbstractAddModuleInfoTask.groovy (hunk 12 above)
Comment (aalmiray) [reply src/main/groovy/org/moditect/gradleplugin/add/AbstractAddModuleInfoTask.groovy:41]: It's also recommended to annotate `File` and `FileCollection` properties with `@PathSensitive` as shown at https://docs.gradle.org/current/userguide/build_cache.html#sec:task_output_caching_details

Comment (siordache) [reply src/main/groovy/org/moditect/gradleplugin/add/AbstractAddModuleInfoTask.groovy:41]: Done.

**Comment by siordache** [comment]
@gunnarmorling: IMO this PR is now ready to be merged.

**Comment by aalmiray** [comment]
+1

**Comment by gunnarmorling** [comment]
Excellent, thanks a lot, @siordache! I'll take another look later this week. It seems also the Vert.x example passes now? Really great stuff! Probably a good idea to get out another release of ModiTect including this plug-in soon.

**Comment by siordache** [comment]
Yes, the Vert.x example passes now.

File: integrationtest/hibernate-validator/build.gradle [hunk 19]
```diff
@@ -0,0 +1,131 @@
+plugins{
```
Comment (gunnarmorling) [inline integrationtest/hibernate-validator/build.gradle:1]: @siordache I know it's noob question, but how do I run this build from the root dir?

**Comment by siordache** [comment]
You can build the plugin and run the tests by executing
`
./gradlew clean build
`

File: src/main/groovy/org/moditect/gradleplugin/aether/DependencyResolver.groovy [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.moditect.gradleplugin.aether
+
+import groovy.transform.CompileStatic
+import org.apache.maven.repository.internal.MavenRepositorySystemUtils
+import org.eclipse.aether.RepositorySystem
```
Comment (gunnarmorling) [inline src/main/groovy/org/moditect/gradleplugin/aether/DependencyResolver.groovy:20]: How comes this is using Aether and not resolution APIs provided by Gradle? AFAIU, resolution rules slightly differ between the two systems, so couldn't there be some surprise in certain cases for a user familiar with the Gradle resolution rules?

File: src/test/groovy/org/moditect/gradleplugin/ModitectPluginSpec.groovy [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+import org.gradle.testkit.runner.BuildResult
+import org.gradle.testkit.runner.GradleRunner
+import org.gradle.testkit.runner.TaskOutcome
+import org.junit.Rule
+import org.junit.rules.TemporaryFolder
+import spock.lang.Specification
+import spock.lang.Unroll
+import spock.util.environment.OperatingSystem
+
+class ModitectPluginSpec extends Specification {
```
Comment (gunnarmorling) [inline src/test/groovy/org/moditect/gradleplugin/ModitectPluginSpec.groovy:27]: Could you explain a bit how this works (again, Gradle noob here)? Will this somehow build/test the integration tests, too?

**Review by gunnarmorling — COMMENTED** [review]
@siordache, wow, this is just awesome; thanks a ton for this! Please see inline for two questions, mostly I'm curious why it's using Aether.

**Comment by gunnarmorling** [comment]
> You can build the plugin and run the tests by executing
./gradlew clean build

Yes, that worked. And where would I then for instance find the modular runtime image for the Undertow example?

**Comment by gunnarmorling** [comment]
I've also filed two issues, #3 and #4, which can be handled separately.

File: src/main/groovy/org/moditect/gradleplugin/aether/DependencyResolver.groovy (hunk 20 above)
Comment (siordache) [reply src/main/groovy/org/moditect/gradleplugin/aether/DependencyResolver.groovy:20]: Optional dependencies are ignored by both Maven and Gradle when building an artifact.
However, Maven internally [keeps track of all dependencies](https://maven.apache.org/ref/3.6.0/maven-core/apidocs/org/apache/maven/project/MavenProject.html#getArtifacts()) and knows [which ones are optional](https://maven.apache.org/ref/3.6.0/maven-artifact/apidocs/org/apache/maven/artifact/Artifact.html#isOptional()).

In Gradle, optional dependencies are completely absent. Aether provides this missing piece of information, but the plugin still works with Gradle's standard data structures (i.e. [Configurations](https://docs.gradle.org/current/dsl/org.gradle.api.artifacts.Configuration.html)).

File: src/test/groovy/org/moditect/gradleplugin/ModitectPluginSpec.groovy (hunk 21 above)
Comment (siordache) [reply src/test/groovy/org/moditect/gradleplugin/ModitectPluginSpec.groovy:27]: In the Spock tests, a [GradleRunner](https://docs.gradle.org/current/javadoc/org/gradle/testkit/runner/GradleRunner.html) is used to build the integration test projects in a temporary folder. The tests check if the build was successful and the launcher files have been created.

For `hibernate-validator`, the Spock test also executes the modular runtime image and checks if the output contains the text "javax.validation.constraints.NotNull.message".

**Comment by siordache** [comment]
> Yes, that worked. And where would I then for instance find the modular runtime image for the Undertow example?


The Spock tests build the runtime images in a temporary folder which is deleted on exit.

After you merge this PR, I will add the plugin to the [Gradle plugin portal](https://plugins.gradle.org/) and then I will provide build scripts for creating modular runtime images of the integration tests from the command line. This way, the runtime images will be created in the _build/image_ directory.

**Comment by siordache** [comment]
@gunnarmorling : When you merge this PR, don't forget to also add the moditect/moditect-gradle-plugin repository to the [moditect-dev](https://github.com/orgs/moditect/teams/moditect-dev/repositories) team.

File: src/main/groovy/org/moditect/gradleplugin/aether/DependencyResolver.groovy (hunk 20 above)
Comment (gunnarmorling) [reply src/main/groovy/org/moditect/gradleplugin/aether/DependencyResolver.groovy:20]: Ok, I see. Thanks for clarifying.

File: src/test/groovy/org/moditect/gradleplugin/ModitectPluginSpec.groovy (hunk 21 above)
Comment (gunnarmorling) [reply src/test/groovy/org/moditect/gradleplugin/ModitectPluginSpec.groovy:27]: Ah, that's pretty cool.

**Review by gunnarmorling — APPROVED** [review]

**Comment by gunnarmorling** [comment]
And it's merged. Huge thank you for this nice work, @siordache!

> I will provide build scripts for creating modular runtime images of the integration tests from the command line. This way, the runtime images will be created in the build/image directory.

That sounds great. Also I was thinking about adding some actual assertions for the Vert.x and Undertow examples by running the images, as it's done for the HV one already. That's very nicely done there.

> I will add the plugin to the Gradle plugin portal

I reckon we need to do a release for that?

> add the moditect/moditect-gradle-plugin repository to the moditect-dev team.

Done.

**Comment by siordache** [comment]
> > I will add the plugin to the Gradle plugin portal

> I reckon we need to do a release for that?

Technically, there is no requirement to keep GitHub releases in sync with releases to the Gradle plugin portal, but I strongly recommend it.

**Comment by siordache** [comment]
@gunnarmorling : Can you please grant access for Travis?
